File: bongf/java/src/dongbinbook/ch18_graph_questions/Q45_FinalRanking.java
package dongbinbook.ch18_graph_questions;

import java.util.*;

// https://www.acmicpc.net/problem/3665
public class Q45_FinalRanking {
    private static Scanner sc;
    public static void main(String[] args) {
        sc = new Scanner(System.in);
        int tc = sc.nextInt();
        for (int i = 0; i < tc; i++) {
            dongbin();
        }
    }

    public static void dongbin() {
        int n = sc.nextInt();
        sc.nextLine();
        int[] indegree = new int[n+1];
        boolean[][] graph = new boolean[n+1][n+1];

        Integer[] prev = Arrays.stream(sc.nextLine().split(" ")).map(Integer::parseInt).toArray(Integer[]::new);
        for (int i = 0; i < n; i++) {
            for (int j = i+1; j < n; j++) {
                graph[prev[i]][prev[j]] = true;
                indegree[prev[j]] += 1;
            }
        }

        int m = sc.nextInt();
        for (int i = 0; i < m; i++) {
            int x = sc.nextInt();
            int y = sc.nextInt();

            if(graph[x][y]) {
                graph[x][y] = false;
                graph[y][x] = true;
                indegree[x]++;
                indegree[y]--;
                continue;
            }

            if(graph[y][x]) {
                graph[y][x] = false;
                graph[x][y] = true;
                indegree[y]++;
                indegree[x]--;
            }
        }

        List<Integer> result = new ArrayList<>();
        Queue<Integer> q = new LinkedList<>();
        for (int i = 1; i < n+1; i++) {
            if(indegree[i] == 0) {
                q.offer(i);
            }
        }

        boolean certain = true;
        boolean cycle = false;

        for (int i = 0; i < n; i++) {
            if(q.isEmpty()) {
                cycle = true;
                break;
            }
            
            if(q.size() >= 2) {
                certain = false;
                break;
            }

            Integer now = q.poll();
            result.add(now);

            for (int j = 1; j <= n; j++) {
                if(graph[now][j]) {
                    indegree[j]--;
                    if(indegree[j] == 0) {
                        q.offer(j);
                    }
                }
            }
        }

        if(cycle) {
            System.out.println("IMPOSSIBLE");
            return;
        }

        if(!certain) {
            System.out.println("?");
            return;
        }

        for (Integer r : result) {
            System.out.print(r + " ");
        }
        System.out.println();
    }
}

File: bongf/python/dongbinbook/ch18_graph_questions/q45_final_ranking.py
from collections import deque

## https://www.acmicpc.net/problem/3665
def dongbin():
    for tc in range(int(input())):
        n = int(input())
        indegree = [0] * (n+1)
        graph = [[False] * (n+1) for _ in range(n+1)]

        data = list(map(int, input().split()))
        for i in range(n):
            for j in range(i+1, n):
                graph[data[i]][data[j]] = True
                indegree[data[j]] += 1
        
        m = int(input())
        for i in range(m):
            x, y = map(int, input().split())
            if graph[x][y]:
                graph[y][x] = True
                graph[x][y] = False
                indegree[x] += 1
                indegree[y] -= 1 
            else:
                graph[x][y] = True
                graph[y][x] = False
                indegree[x] -= 1
                indegree[y] += 1
        
        result = []
        q = deque()

        for i in range(1, n+1):
            if indegree[i] == 0:
                q.append(i)
        
        certain = True
        cycle = False

        for i in range(n):
            if len(q) == 0: 
                cycle = True
                break
                
            if len(q) >= 2 :
                certain = False
                break
            
            now = q.popleft()
            result.append(now)

            for j in range(1, n+1):
                if graph[now][j]:
                    indegree[j] -= 1
                    if indegree[j] == 0 :
                        q.append(j)
        
        if cycle: 
            print("IMPOSSIBLE")
        elif not certain:
            print("?")
        else:
            for i in result:
                print(i, end=" ")
            print()

dongbin()
